### PR TITLE
Add upload progress + headers

### PR DESCRIPTION
**Pull request by PVince81**
Requires https://github.com/owncloud/davclient.js/pull/39/

### CODE REVIEW

**Comment by PVince81** [comment]
- [ ] TODO: apply the same for public links

or refactor the whole thing to use only one code path for uploads...

File: package.json [hunk 1]
```diff
@@ -31,8 +31,10 @@
     "lint-fix": "eslint tests/**/*.js tests/*.js src/**/*.js src/*.js --color --global requirejs --global require --fix",
     "test": "karma start --single-run --browsers ChromeHeadless",
     "test-drone": "karma start --single-run --browsers ChromeHeadlessNoSandbox",
+    "prepare": "npm run build:dev",
```
Comment (DeepDiver1975) [inline package.json]: Is that needed within the final PR in here? I'm aware of the local dev reasons for this - I'm wondering about sideeffects on the real build/publish - THX

Comment (PVince81) [reply package.json]: I hope we can keep it including the source map thing, especially if we're going to do more development on the SDK in the near future.

But I have the same worry like yours. I'll research this

**Comment by PVince81** [comment]
- [ ] research whether "prepare" is skipped for production builds

**Comment by PVince81** [comment]
According to https://docs.npmjs.com/misc/scripts it seems "prepare" is only using when publishing packages and when installing git dependencies.

But then it means that if I put the source map there (the dev build) it would also package that in the production version.

Doesn't seem like a good solution so I'm fine removing these bits before merging.

**Comment by PVince81** [comment]
ref where I found out about npm prepare: https://stackoverflow.com/questions/48287776/automatically-build-npm-module-on-install-from-github

File: package.json (hunk 1 above)
Comment (DeepDiver1975) [reply package.json]: We could also skip webpack totally - I did add webpack to get transpiling in place.
With the knowledge of today: any consuming project can transpile any dependency to it's own needs.

the sdk can be shippes as pure sources ... which would also make debugging on the consuming project (e.g. phoenix) easier

Comment (PVince81) [reply package.json]: if we use ES6, would transpiling be needed for running the tests in the browser ?

Comment (DeepDiver1975) [reply package.json]: tests have anyhow their own setup/config - this is unrelated to the way we distribute the package

**Comment by DeepDiver1975** [comment]
In order to get this rolling I suggest to split the PR: upload progress and build setup

**Comment by PVince81** [comment]
added code for public link
added onProgress assertion in the tests (and used the opportunity to use async/await there)

**Comment by PVince81** [comment]
finally got tests passing.

next up: cleanup

**Comment by PVince81** [comment]
@DeepDiver1975 I've removed the "dev mode" stuff, but moved it to a separate branch for later reconsideration.

@DeepDiver1975 please review, this is ready to ship

**Comment by DeepDiver1975** [comment]
@PVince81 there is something wrong in the area of public files - see ci ....

**Comment by PVince81** [comment]
it is possible that for small files `onProgress` is never called... the spec doesn't tell how often a browser is supposed to call it, it just says "periodically".

maybe for this one test we should mock the davclient...

**Comment by PVince81** [comment]
all tests pass for me locally :frowning:

**Comment by PVince81** [comment]
I've found this, so it could be timing related: https://stackoverflow.com/a/5543745

I'll see if it's easier to use a mock instead

**Comment by PVince81** [comment]
- [ ] create new folder "unit" for unit tests using mocks when needed, to make sure there is a clear separation

**Comment by PVince81** [comment]
seems Sinon JS have changed their APIs and I can't find the progress object https://sinonjs.org/releases/latest/fake-xhr-and-server/
also no luck with Jasmine's ajax XHR mock which doesn't have the "upload" attribute https://github.com/jasmine/jasmine-ajax/issues/109

**Comment by PVince81** [comment]
as an alternative attempt, I tried uploading 1 MB+ content but the progress events still don't fire, despite the stack overflow article there. I wonder if this is related to ChromeHeadless that progress events are not fired

**Comment by PVince81** [comment]
another hope here that would bring us closer to how browsers actually upload: https://stackoverflow.com/a/43767079

- [ ] try uploading with FormData or other methods like binary blob, as they are more likely to trigger progress events

**Comment by PVince81** [comment]
Uploading with FormData produces extra contents, but not usable.

- [x] try with Blob

Unfortunately uploading a Blob directly also does not trigger progress events :rage:

**Comment by PVince81** [comment]
Big facepalm!

The yarn.lock was still pointing at the old davclient.js that had no onProgress event.
I didn't think of going to debug that deep...

Updated now, tests should pass!

**Comment by PVince81** [comment]
@DeepDiver1975 this is ready for re-review and merging!

**Pull request merged**